spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -534,39 +541,45 @@ function chanUpgradeTry(
     // construct counterpartyChannel from existing information and provided
     // counterpartyUpgradeSequence
     counterpartyChannel = ChannelEnd{
-        state: INITUPGRADE,
+        state: OPEN,
         ordering: currentChannel.ordering,
         counterpartyPortIdentifier: portIdentifier,
         counterpartyChannelIdentifier: channelIdentifier,
         connectionHops: counterpartyHops,
         version: currentChannel.version,
         sequence: counterpartyUpgradeSequence,
-        flushStatus: NOTINFLUSH,
     }
 
-    // call startFlushUpgrade handshake to move channel from INITUPGRADE to TRYUPGRADE and start flushing
-    // upgrade is blocked on this channelEnd from progressing until flush completes on both ends
-    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyChannel, counterpartyUpgrade, TRYUPGRADE, proofChannel, proofUpgrade, proofHeight)
+    // call startFlushUpgrade handshake to move channel to FLUSHING, which will block
+    // upgrade from progressing to OPEN until flush completes on both ends
+    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyChannel, counterpartyUpgrade, proofChannel, proofUpgrade, proofHeight)
 
     // refresh currentChannel to get latest state
     currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
 
     // call modules onChanUpgradeTry callback
     module = lookupModule(portIdentifier)
     version, err = module.onChanUpgradeTry(
-        proposedUpgradeChannel.ordering,
-        proposedUpgradeChannel.connectionHops,
         portIdentifier,
         channelIdentifer,
+        proposedUpgradeChannel.ordering,
+        proposedUpgradeChannel.connectionHops,
         currentChannel.sequence,
         proposedUpgradeChannel.counterpartyPortIdentifer,
         proposedUpgradeChannel.counterpartyChannelIdentifier,
         proposedUpgradeChannel.version
     )
-    // restore channel if callback returned error
-    if err != nil {
+    // abort the transaction if the callback returns an error and
+    // there was no existing upgrade. This will allow the counterparty upgrade
+    // to continue existing while this chain may add support for it in the future
+    if err != nil && existingUpgrade == nil {
+        abortTransactionUnless(false)
+    }
+    // if the callback returns an error while an existing upgrade is in place
+    // or if the existing upgrade is not compatible with the counterparty upgrade
+    // we must restore the channel so that a new upgrade attempt can be made
+    if err != nil || existingUpgrade.fields != upgradeFields {
         restoreChannel(portIdentifier, channelIdentifier)
```

Comment:
probably can become:
```suggestion
    // abort the transaction if the callback returns an error and
    // This will allow the counterparty upgrade to continue existing while this chain may add support for it in the future
    if err != nil {
        abortTransactionUnless(false)
    }
```
if you use existing upgrade